enterprise-suite/templates/es-console.yaml

```diff
@@ -72,22 +79,111 @@ data:
+      #   magic 'replace' value ' $1/' with a space prefix tells nginx to not add server and port
+      # sub_filter 'match' 'replace'
+      #   rewrites backend content, which we use for inserting the base path
+      # proxie_cookie_path 'match' 'replace'
+      #   fixes session cookie Path attribute, for Grafana
+
+      # proxy_pass backends with dynamic base url
+
+      sub_filter_once off;
+
+      location ~ (.*/service/prometheus)(/.*) {
+        proxy_pass http://$prometheus$2$is_args$args;
+        sub_filter '="/' '="$1/';
+        sub_filter 'PATH_PREFIX = ""' 'PATH_PREFIX = "$1"';
+        proxy_redirect '/' ' $1/';
       }
 
-      location /service/prometheus/ {
-        proxy_pass http://prometheus-server/;
+      location ~ (.*/service/es-monitor-api)(/.*) {
+        proxy_pass http://$monitorapi$2$is_args$args;
+        proxy_redirect '/' ' $1/';
       }
 
-      location /service/es-monitor-api/ {
-        proxy_pass http://es-monitor-api/;
+      location ~ (.*/service/grafana)(/.*) {
+        proxy_pass http://$grafana:3000$2$is_args$args;
+        sub_filter_types *;
+        sub_filter '/service/grafana' '$1';
+        proxy_cookie_path '/service/grafana' '$1';
+        proxy_redirect '/' ' $1/';
       }
 
-      location /service/grafana/ {
-        proxy_pass http://grafana-server:3000/;
+      # our injected plugin should never cache...
+      location ~ (.*/service/grafana)(/dashboard/script/exporter-async.js) {
+        proxy_pass http://$grafana:3000$2$is_args$args;
+        sub_filter_types *;
+        sub_filter '/service/grafana' '$1';
+        proxy_cookie_path '/service/grafana' '$1';
+        proxy_redirect '/' ' $1/';
+        # no-cache
+        add_header Last-Modified $date_gmt;
+        add_header Cache-Control 'no-store, no-cache, must-revalidate, proxy-revalidate, max-age=0';
+        if_modified_since off;
+        expires off;
+        etag off;
       }
+
+      # redirect /service/prometheus -> /service/prometheus/
+
+      location ~ (.*/service/(prometheus|grafana))$ {
+        return 301 ' $1/';
+      }
+
+      # console entry point
+
+      location ~ (.*/)index.html {
+        alias $base/index.html;
+        sub_filter 'base href="/"' 'base href="$1"';
+        # no-cache
+        add_header Last-Modified $date_gmt;
+        add_header Cache-Control 'no-store, no-cache, must-revalidate, proxy-revalidate, max-age=0';
+        if_modified_since off;
+        expires off;
+        etag off;
+      }
+
+      # bookmarks / reloads in the /workloads/ path
+      # (for our single page app, these should reload index.html but keep the path)
+
+      location ~ (.*/)(workloads/.*) {
+        alias $base/index.html;
+        default_type text/html;
+        sub_filter 'base href="/"' 'base href="$1"';
+        # kill cache
+        add_header Last-Modified $date_gmt;
+        add_header Cache-Control 'no-store, no-cache, must-revalidate, proxy-revalidate, max-age=0';
```

Comment:
For the line `add_header Cache-Control 'no-store, no-cache, must-revalidate, proxy-revalidate, max-age=0'` , we duplicated it in 3 different places.  Could we set a variable for it?

Reply:
ya nginx doesn't seem to have very good facilities for composing configuration, but we can certainly define a common string.  I couldn't find a good way to have the whole section common, so there is some copy paste